hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/NameNodeRpcServer.java

```diff
@@ -1100,20 +1100,29 @@ public void rename2(String src, String dst, Options.Rename... options)
   }
 
   @Override // ClientProtocol
-  public boolean truncate(String src, long newLength, String clientName)
-      throws IOException {
+  public boolean truncate(String src, long newLength, String clientName) throws IOException {
     checkNNStartup();
-    stateChangeLog
-        .debug("*DIR* NameNode.truncate: " + src + " to " + newLength);
+    if(stateChangeLog.isDebugEnabled()) {
+      stateChangeLog.debug("*DIR* NameNode.truncate: " + src + " to " +
+          newLength);
+    }
+    CacheEntryWithPayload cacheEntry = RetryCache.waitForCompletion(retryCache, null);
+    if (cacheEntry != null && cacheEntry.isSuccess()) {
+      return (boolean)cacheEntry.getPayload();
+    }
+
     String clientMachine = getClientMachine();
+    boolean ret = false;
     try {
-      return namesystem.truncate(
+      ret = namesystem.truncate(
           src, newLength, clientName, clientMachine, now());
     } finally {
+      RetryCache.setState(cacheEntry, true, ret);
```

Comment:
Finally block will be executed in case of exception as well, we can not hard-code `true` here. 
Can check other codes like for `append` to get some reference & idea.